### PR TITLE
Fix array equal check

### LINKED ISSUE #570: ListArray equality check may return wrong result
**Describe the bug**
The `PartialEq` impl on `ListArray` will ignore some differences.

**To Reproduce**

[This case](https://github.com/waynexia/arrow-rs/blob/7f6ce9299bc3a51665a26ef5da788f1e627e9be5/arrow/src/compute/kernels/sort.rs#L2421-L2436) should fail but not.

**Expected behavior**
The above case fail.

**Additional context**
I noticed this from [this test case](https://github.com/waynexia/arrow-rs/blob/7f6ce9299bc3a51665a26ef5da788f1e627e9be5/arrow/src/compute/kernels/sort.rs#L2357-L2366) in the master branch, which is wrong but can pass.

I think the root cause is when comparing two list arrays, the [children's bitmap](https://github.com/waynexia/arrow-rs/blob/7f6ce9299bc3a51665a26ef5da788f1e627e9be5/arrow/src/array/equal/list.rs#L130-L134) were miscalculated leaving the remaining procedure to recognize some differences as null.

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #570.

# What changes are included in this PR?

This PR fixes the problem mentioned in #570 and adds some negative test cases.

# Are there any user-facing changes?
No.

### CODE REVIEW

**Review by jorgecarleitao — APPROVED** [review]
Thanks @waynexia , looks great 👍

File: arrow/src/compute/kernels/sort.rs [hunk 1]
```diff
@@ -2347,7 +2347,7 @@ mod tests {
                 nulls_first: true,
             }),
             Some(3),
-            vec![None, None, Some(vec![Some(2)])],
+            vec![None, None, Some(vec![Some(1)])],
```
Comment (alamb) [inline arrow/src/compute/kernels/sort.rs:2350]: 👍

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
FYI @nevi-me  and @bjchambers

**Comment by alamb** [comment]
Thanks again @waynexia

**Pull request merged**